pkg/certs/certificateAuthority.go

```diff
@@ -0,0 +1,301 @@
+		Intermediates: intermediatesPool,
+		Roots:         ca.rootsPool,
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+	}
+
+	_, err := leaf.Verify(verifyOptions)
+
+	if err != nil {
+		return fmt.Errorf("unable to verify client certificate: %v", err)
+	}
+
+	return nil
+}
+
+func (ca *CertificateAuthority) NewSignedCert(csr *x509.CertificateRequest, conf SignConfig) (*x509.Certificate, error) {
+	var err error
+	err = csr.CheckSignature()
+	if err != nil {
+		return nil, fmt.Errorf("invalid CSR signature: %v", err)
+	}
+
+	serial, err := rand.Int(rand.Reader, new(big.Int).SetInt64(math.MaxInt64))
+	if err != nil {
+		return nil, err
+	}
+
+	now := time.Now().UTC()
+
+	tmpl := x509.Certificate{
+		SerialNumber:          serial,
+		Subject:               csr.Subject,
+		NotBefore:             now,
+		NotAfter:              now.Add(conf.offset), // 1 year
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+		DNSNames:              csr.DNSNames,
+		IPAddresses:           csr.IPAddresses,
+	}
+
+	b, err := x509.CreateCertificate(rand.Reader, &tmpl, ca.rootCert, csr.PublicKey, ca.rootSigner.PrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return x509.ParseCertificate(b)
+}
+
+func (ca *CertificateAuthority) SignRenewRequest(csr *x509.CertificateRequest, oldCert *x509.Certificate) (*x509.Certificate, error) {
+	var err error
+	err = csr.CheckSignature()
+	if err != nil {
+		return nil, fmt.Errorf("invalid CSR signature: %v", err)
+	}
+
+	csrRenewCertsPEM := []byte{}
+
+	for _, ext := range csr.Extensions {
+		if ext.Id.Equal(oidRenewCertificates) {
+			csrRenewCertsPEM = ext.Value
+			break
+		}
+	}
+
+	if len(csrRenewCertsPEM) == 0 {
+		return nil, fmt.Errorf("missing CSR renew certificates extension")
+	}
+
+	csrRenewCert, err := parseCertsPEM(csrRenewCertsPEM)
+	if err != nil || len(csrRenewCert) < 2 {
+		return nil, fmt.Errorf("missing CSR renew certificates")
+	}
+
+	// csrRenewCert[0] is signed by csrRenewCert[1]
+	// csrRenewCert[1] is cert to be renewed
+	// csrRenewCert[2] ... optional intermediate certificates to verify csrRenewCert
+
+	certToRenew := csrRenewCert[1]
+
+	// The certToRenew must also be used as the clientAuthCert
+	if oldCert != nil {
+		if !bytes.Equal(certToRenew.Raw, oldCert.Raw) {
+			return nil, fmt.Errorf("certToRenew wasn't used for clientAuthCert")
+		}
+	}
+
+	intermediatesPool := x509.NewCertPool()
+	if len(csrRenewCert) > 2 {
+		for _, cert := range csrRenewCert[2:] {
+			intermediatesPool.AddCert(cert)
+		}
+	}
+
+	verifyOptions := x509.VerifyOptions{
+		Intermediates: intermediatesPool,
+		Roots:         ca.rootsPool,
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+	}
+
+	_, err = certToRenew.Verify(verifyOptions)
+	if err != nil {
+		if time.Now().After(certToRenew.NotAfter) {
+			return nil, fmt.Errorf("unable to verify certificate to be renewed: Certificate Expired: %v", err)
```

Comment:
You may want to consider defining an error in our "errors.go" module (https://github.com/microsoft/moc/blob/main/pkg/errors/errors.go), because "certificate expired" is definitely something the calling code can automatically act on (i.e. the calling code may be able to update the certificate).

Reply:
Once you have defined your error, feel free to use errors.Wrapf(<your error>, "unable to verify certificate to be renewed: Certificate Expired: %v", err)

Reply:
Same comment applies to locations below. I think some of these (if not all) may be recoverable by the calling code.

Reply:
Sure, It has another revision of the changes. I incorporate the suggested changes into that. Thanks

Reply:
Thanks a lot! Looks good!